certora/WormholeRouter.spec

```diff
@@ -0,0 +1,251 @@
+    uint256 wardOther = wards(other);
+
+    rely(e, usr);
+
+    assert(wards(usr) == 1, "rely did not set the wards as expected");
+    assert(wards(other) == wardOther, "rely affected other wards which wasn't expected");
+}
+
+// Verify revert rules on rely
+rule rely_revert(address usr) {
+    env e;
+
+    uint256 ward = wards(e.msg.sender);
+
+    rely@withrevert(e, usr);
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = ward != 1;
+
+    assert(revert1 => lastReverted, "revert1 failed");
+    assert(revert2 => lastReverted, "revert2 failed");
+
+    assert(lastReverted => revert1 || revert2, "Revert rules are not covering all the cases");
+}
+
+// Verify that wards behaves correctly on deny
+rule deny(address usr) {
+    env e;
+
+    address other;
+    require(other != usr);
+    uint256 wardOther = wards(other);
+
+    deny(e, usr);
+
+    assert(wards(usr) == 0, "deny did not set the wards as expected");
+    assert(wards(other) == wardOther, "deny affected other wards which wasn't expected");
+}
+
+// Verify revert rules on deny
+rule deny_revert(address usr) {
+    env e;
+
+    uint256 ward = wards(e.msg.sender);
+
+    deny@withrevert(e, usr);
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = ward != 1;
+
+    assert(revert1 => lastReverted, "revert1 failed");
+    assert(revert2 => lastReverted, "revert2 failed");
+
+    assert(lastReverted => revert1 || revert2, "Revert rules are not covering all the cases");
+}
+
+// Verify that gateway behaves correctly on file
+rule file_domain_address(bytes32 what, bytes32 domain, address data) {
+    env e;
+
+    bool whatIsGateway = what == 0x6761746577617900000000000000000000000000000000000000000000000000;
+    bool dataIsEmpty = data == 0x0000000000000000000000000000000000000000;
+    address gatewayBefore = gateways(domain);
+    bool gatewayWasEmpty = gatewayBefore == 0x0000000000000000000000000000000000000000;
+    uint256 numDomainsBefore = numDomains();
+    bool hasDomainBefore = hasDomain(domain);
+
+    file(e, what, domain, data);
+
+    uint256 numDomainsAfter = numDomains();
+
+    assert(
+        whatIsGateway => gateways(domain) == data, "file did not set gateways(domain) as expected"
+    );
+    assert(
+        whatIsGateway && !dataIsEmpty
+        => domains(data) == domain, "file did not set domains(gateway) as expected"
+    );
+    assert(
+        whatIsGateway && gatewayWasEmpty && !dataIsEmpty && !hasDomainBefore && numDomainsBefore < max_uint256
+        => numDomainsAfter == numDomainsBefore + 1, "file did not increase allDomains length as expected");
+    assert(
+        whatIsGateway && gatewayWasEmpty && !dataIsEmpty
+        => domainAt(numDomainsBefore) == domain, "file did not set allDomains as expected");
+    assert(
+        whatIsGateway && !gatewayWasEmpty
+        => domains(gatewayBefore) == 0x0000000000000000000000000000000000000000000000000000000000000000, "file did not set domains(gateway) as expected 2");
```

Comment:
I am extremely confused as to how the tool does not find a counterexample to this. I confirmed the following test, when added to WormholeRouter.t.sol, passes:

```
    function testFileSameGatewayForDomain() public {
        bytes32 domain = "dom";

        address gateway1 = address(111);
        assertTrue(_tryFile("gateway", domain, gateway1));
        assertEq(router.gateways(domain), gateway1);
        assertEq(router.domains(gateway1), domain);
        assertEq(router.numDomains(), 1);
        assertEq(router.domainAt(0), domain);

        // re-file the same gateway for the same domain
        assertTrue(_tryFile("gateway", domain, gateway1));
        assertEq(router.domains(gateway1), domain);
    }
```

The second call to `file` there should satisfy the following three conditions:
1) `what` is "gateway"
2) `gateways(domain)` is non-empty
3) after the second call, `domains(gatewayBefore)` is non-empty because after the call it's still "dom" as we just re-filed the same gateway for the same domain.

That's a counterexample to this assertion, right? Or did I take too many crazy pills again?

Reply:
I'll add that I don't see any requirement in this rule that `data != gatewayBefore`, which would be one way to avoid finding a counterexample here.

Reply:
Fixed in: https://github.com/makerdao/dss-wormhole/pull/40/commits/3d4960e708800afcedbd6f56065607ffaf360b3c